### ISSUE TITLE
x509 failure handling

### ISSUE THREAD

**Issue by hramrach** [ipxe/ipxe#1069]
Hello,

I have a server from which I cannot boot using iPXE due to x509 failure.

Building with DEBUG=x509 I get certificate dump on the VGA console of which I see only the final lines which is not immensely helpful.

After some discussion on irc the probable cause was identified as one of the intermediate certificated in the chain setting 'Certificate Name Constraints' as critical extension.

**Comment by NiKiZe** [comment]
The message you get (the original before you enabled debugging) is short, but does include an url to ipxe.org with information and things to test, including steps.
I have personally found that using `DEBUG=validator` is better at showing the offending underlaying issue.

If you already have a syslog server setup, using `set syslog ip.v4.add.ress` is a great way (one of several) to grab the full output.

**Comment by hramrach** [comment]
I could not determine the reason for the failure from the information provided by iPXE, hence the bug.

**Comment by NiKiZe** [comment]
> I could not determine the reason for the failure from the information provided by iPXE, hence the bug.

Understood, there probably is things in documentation that could be improved, but to do that that it is helpful to have the link. ;)
I read the issue as request for help, but without any information that is needed to actually help.

Now looking in to this, and something that might warrant an issue is `Certificate Name Constraints (OID 2.5.29.30)` , which "Per RFC 5280 Section 4.2.1.10, the extension should be marked as critical", but if a Client don't support a node marked as critical it must reject the certificate.
So we could open an issue for iPXE to support id-ce-nameConstraints if this is still an issue for you?

**Comment by hramrach** [comment]
It's been a while and the network configuration has hanged. I do not use the particular configuration anymore.

The problem at the time was that the reason for rejecting the certificate was not clear.

**Comment by NiKiZe** [comment]
There is only so much info that can be supplied in the limited scope in code, often only "permission denied" or similar will be what you get, to get the details you need to enable debug. And if possible we will improve the documentation the error links to and explain how to enable and capture the information needed to determine the actual cause.

**Comment by hramrach** [comment]
As already said enabling debug is not helpful. It provides a lot of information but on a VGA console with no scrollback most of it is lost.

**Comment by NiKiZe** [comment]
Agree that using `DEBUG=x509` is not the best option here, but using other ways of grabing the data or `DEBUG=validator` or `=tls` is.

**Comment by hramrach** [comment]
So if it's well documented what is likely to provide the relevant information and cut out the irrelevant this would be likely resolved.

**Comment by mcb30** [comment]
The error message would have already pointed you to https://ipxe.org/err/3c16e2 which already tells you that the problem is due to an unsupported extension, and includes a direct link to the single line of code capable of generating that error, where you can see immediately from the comments that the problem is that a certificate has an unsupported critical extension: https://github.com/ipxe/ipxe/blob/94138656d74a5f787ca6c6dd22c08e656ac131f3/src/crypto/x509.c#L878-L883

That seems like quite a lot of high precision and directly relevant information already.  I'm not sure what more you could reasonably expect to be given.....?

**Comment by hramrach** [comment]
The information given is
 - read source code (who other than iPXE developers understands that?)
 - Try using the [latest version](https://ipxe.org/download) of iPXE. Your problem may have already been fixed.
 - Try building iPXE with the [debug](https://ipxe.org/download#debug_builds) option DEBUG=x509
 - You can [contact](https://ipxe.org/contact) the iPXE developers and other iPXE users.
 - [Refresh](https://ipxe.org/err/3c16e2) this page after 24 hours. This page is actively monitored, and further information may be added soon.

That does not lead to diagnosing the problem, at least not in a reasonable time.

In particular it's unclear what extension was included and not handled.

**Comment by NiKiZe** [comment]
Oh, seems I updated that page recently due to creating a certificate myself with non supported extension marked as critical. If just the url, as shown in the original error message, even before debug being enabled, would have been included, it would have been easy.

Sorry but now you are **intentionally**? not reading that page, but yes you will have to enable debug to collect exactly which critical section isn't supported, there are other ways than a EGA size console to view that output.

**Comment by mcb30** [comment]
> The information given is
>     * read source code (who other than iPXE developers understands that?)

There is also the rather important part of the information given where it says, in bold, at the top of the page: "**Error: Unsupported extension**".

And no, it doesn't just say "read source code".  It gives you a link straight to the **precise line of the relevant file** in the source code, where even someone who doesn't write C can see the comment `Fail if we cannot handle a critical extension` written immediately above it.

**Comment by hramrach** [comment]
Enabling debug as instructed provides no information because the information does not fit on the screen, as already reported.

The problem was diagnosed to be with intermediate certificate, not the server certificate. That means examining the server certificate will likely NOT show the extenstion that iPXE complains about.

Also if I as a user get unsupported extension what is the way forward?

Probably stop using iPXE if I am not willing to write the support if I even get through all the confusion to conclude that this is a feature missing in iPXE.

**Comment by hramrach** [comment]
And no, I don't know what /* Fail if we cannot handle a critical extension */ means as a non-iPXE developer. That's not user documentation.

**Comment by NiKiZe** [comment]
> Enabling debug as instructed provides no information because the information does not fit on the screen, as already reported.

Already tried to give suggestion on how to get the data, error page updated to include what was already suggested here, but also other ways.

> 
> The problem was diagnosed to be with intermediate certificate, not the server certificate. That means examining the server certificate will likely NOT show the extenstion that iPXE complains about.

Yes the full chain needs to be investigated, did it say to **only** check the server cert?

> 
> Also if I as a user get unsupported extension what is the way forward?
> 
> Probably stop using iPXE if I am not willing to write the support if I even get through all the confusion to conclude that this is a feature missing in iPXE.

Lets read what the page currently says again (fully aware of that this wasn't the case when the issue was created):
> Certificate have an extension that is marked as critical that iPXE don't support. Consider if this extension should be critical and if not change it and reissue certificate. If it really needs to be critical then some development will be needed in iPXE, the certificate or at least extension needs to be sent to a developer. (You need to enabled DEBUG as described above to see exactly what fails)

So:
### Consider if this extension should be critical and if not change it and reissue certificate.

To use iPXE, or any client, the server needs to be configured in a compatible way, this includes cipers, **and certificates**.
Or have the wanted features being developed, as I suggested earlier, open an issue that state requested feature.

One example here is letsencrypt which gave certificates that was not compatible, re-issue was needed.

There is even the option to run a fully self signed chain, so self signed ca certificate and cross-sign the problematic certificate with the issue, and have iPXE working with that as embeded/trusted chain.


The now updated error page should cover the initial difficulties around this issue.

**Comment by mcb30** [comment]
> Enabling debug as instructed provides no information because the information does not fit on the screen, as already reported.

Yes, debug messages can be verbose.  Start with the more immediately available information: the error message "Unsupported extension" is alone enough to point you in the right direction of the problem.

> The problem was diagnosed to be with intermediate certificate, not the server certificate. That means examining the server certificate will likely NOT show the extenstion that iPXE complains about.

Correct.  All of the certificates in a chain have to be validated.  This is just how X.509 certificates work.  It's not iPXE-specific in any way.

> Also if I as a user get unsupported extension what is the way forward?

Reissue the offending certificate without using that relatively niche certificate extension.

> And no, I don't know what /* Fail if we cannot handle a critical extension */ means as a non-iPXE developer. That's not user documentation.

No, it's a code comment.  If you have sufficient expertise to rebuild the code with a debug option enabled, then you definitely have enough expertise to read a code comment.

**Comment by hramrach** [comment]
> No, it's a code comment. If you have sufficient expertise to rebuild the code with a debug option enabled, then you definitely have enough expertise to read a code comment.

Yes, I can read the code comment. It does not give the information needed for somebody not familiar with iPXE x509 internals.

> Reissue the offending certificate without using that relatively niche certificate extension."

Not my certificate, not possible.

> Already tried to give suggestion on how to get the data, error page updated to include what was already suggested here, but also other ways.

Nope, it still says to enable x509 debug which fails to provide the useful information,

> Yes the full chain needs to be investigated, did it say to only check the server cert?

It now mentions the chain so at least this is probably resolved.

**Comment by hramrach** [comment]
> Yes, debug messages can be verbose. Start with the more immediately available information: the error message "Unsupported extension" is alone enough to point you in the right direction of the problem.

Amy be for you. For a user of the software absolutely not.

**Comment by mcb30** [comment]
> > Already tried to give suggestion on how to get the data, error page updated to include what was already suggested here, but also other ways.
> 
> Nope, it still says to enable x509 debug which fails to provide the useful information,

Actually, the `DEBUG=x509` output will have included the message:
```
X509 0x.... cannot handle critical extension:
<hex dump of the unsupported critical extension data>
```
which is *precisely* the information you need, if you are trying to diagnose the problem *by looking at the debug output* instead of using any of the other sources of information that were made available to you.

At this point, you are complaining that you have been provided with all of:
* a meaningful error title "Unsupported extension"
* a debug message telling you exactly what went wrong (along with plenty of other debug information)
* a direct link to the relevant line of source code that also tells you exactly what went wrong 
* a precise error message that would allow anyone familiar with the code to have told you exactly what the problem was, if only you'd bothered to report what the error message actually was

but that this is still somehow not helpful enough for you.

Given this, could you please write up a document that outlines all of the errors that could possibly arise when validating a certificate chain, and we will make sure to direct future users to it.  (A quick analysis finds around 250 possible error cases that you will need to document.)

**Comment by NiKiZe** [comment]
As I understand it the complaint is that there is to much output to fit on your small display.
Well the relevant output is there, and If all of it is wanted I have explained how to get that, and that is in the update.
A screenshot of what you had would have been enough.

For the sake of posterity:
> Certificate in chain have an extension that is marked as critical that iPXE don't support.
Consider if this extension should be critical and if not change it and reissue certificate.
If it really needs to be critical then some development will be needed in iPXE, the certificate or at least extension needs to be sent to a developer. (You need to enabled DEBUG as described above to see exactly what fails)

> Since the output can be quite verbose consider sending this to `syslog` or serial.
Another option is to run the linux binary of iPXE which usually gives you a friendlier interface to get text from:
```bash
        make -C src bin-x86_64-linux/ipxe.linux DEBUG=x509
        src/bin-x86_64-linux/ipxe.linux --net slirp
```

> To decode your certificate you can use services such as https://asn1js.eu/
